cli/printlog/compact/compact.go

```diff
@@ -0,0 +1,267 @@
+type SpawnLogReconstructor struct {
+	input *bufio.Reader
+
+	hashFunc string
+	files    map[int32]*spb.File
+	dirs     map[int32]*reconstructedDir
+	symlinks map[int32]*spb.File
+	sets     map[int32]*spb.ExecLogEntry_InputSet
+
+	sort bool
+}
+
+type reconstructedDir struct {
+	path  string
+	files []*spb.File
+}
+
+func NewSpawnLogReconstructor(input io.Reader, sort bool) *SpawnLogReconstructor {
+	return &SpawnLogReconstructor{
+		input:    bufio.NewReader(input),
+		hashFunc: "",
+		files:    make(map[int32]*spb.File),
+		dirs:     make(map[int32]*reconstructedDir),
+		symlinks: make(map[int32]*spb.File),
+		sets:     make(map[int32]*spb.ExecLogEntry_InputSet),
+		sort:     sort,
+	}
+}
+
+func (slr *SpawnLogReconstructor) GetSpawnExec() (*spb.SpawnExec, error) {
+	entry := &spb.ExecLogEntry{}
+	for {
+		err := protodelim.UnmarshalFrom(slr.input, entry)
+		if err == io.EOF {
+			return nil, io.EOF
+		}
+		if err != nil {
+			return nil, fmt.Errorf("failed to read execution log entry: %s", err)
+		}
+
+		switch e := entry.GetType().(type) {
+		case *spb.ExecLogEntry_Invocation_:
+			slr.hashFunc = e.Invocation.GetHashFunctionName()
+		case *spb.ExecLogEntry_File_:
+			slr.files[entry.GetId()] = reconstructFile(nil, e.File)
+		case *spb.ExecLogEntry_Directory_:
+			slr.dirs[entry.GetId()] = reconstructDir(e.Directory)
+		case *spb.ExecLogEntry_UnresolvedSymlink_:
+			slr.symlinks[entry.GetId()] = reconstructSymlink(e.UnresolvedSymlink)
+		case *spb.ExecLogEntry_InputSet_:
+			slr.sets[entry.GetId()] = e.InputSet
+		case *spb.ExecLogEntry_Spawn_:
+			return slr.reconstructSpawn(e.Spawn)
+		default:
+			log.Warnf("unknown exec log entry: %v", entry)
+		}
+	}
+}
+
+func (slr *SpawnLogReconstructor) reconstructSpawn(s *spb.ExecLogEntry_Spawn) (*spb.SpawnExec, error) {
+	se := &spb.SpawnExec{
+		CommandArgs:     s.GetArgs(),
+		TargetLabel:     s.GetTargetLabel(),
+		Mnemonic:        s.GetMnemonic(),
+		ExitCode:        s.GetExitCode(),
+		Status:          s.GetStatus(),
+		Runner:          s.GetRunner(),
+		CacheHit:        s.GetCacheHit(),
+		Remotable:       s.GetRemotable(),
+		Cacheable:       s.GetCacheable(),
+		RemoteCacheable: s.GetRemoteCacheable(),
+		TimeoutMillis:   s.GetTimeoutMillis(),
+		Metrics:         s.GetMetrics(),
+		Digest:          s.GetDigest(),
+	}
+	// Handle env vars
+	envVar := s.GetEnvVars()
+	if slr.sort {
+		slices.SortFunc(envVar, compMapEntries)
+	}
+	se.EnvironmentVariables = envVar
+
+	// Handle platform
+	platform := s.GetPlatform()
+	if slr.sort && platform != nil {
+		slices.SortFunc(platform.Properties, compMapEntries)
+	}
+	se.Platform = platform
+
+	// Handle inputs
+	inputs := slr.reconstructInputs(s.GetInputSetId())
+	toolInputs := slr.reconstructInputs(s.GetToolSetId())
+	var spawnInputs []*spb.File
+	for path, file := range inputs {
+		if _, ok := toolInputs[path]; ok {
+			file.IsTool = true
+		}
+		spawnInputs = append(spawnInputs, file)
+	}
+	if slr.sort {
```

Comment:
The old exec format didn't sort the individual fields of the spawn proto as those are already deterministically ordered (based on `depset` order). Instead, to make the log more comparable overall, we would need to sort the protos with respect to each other. 

Bazel topologically sorts by consumer-producer relationship and then by output path, see https://github.com/bazelbuild/bazel/blob/5d4feefed7e39b20a6c5deb5f74394abbf622a52/src/main/java/com/google/devtools/build/lib/exec/StableSort.java#L55. I think that this is what we should do if `-sort` is specified.

Reply:
Im a bit afraid of implementing this.

Currently, we are converting from the new compact format to the old format, which is big and could take a long time to sort.
But perhaps it's fine because we are not processing the log inside the invocation. 🤔 

Let me give it a try.

Reply:
I don't think it's possible to sort the new compact format directly as consumer/producer relationships can only be discovered after you have "flattened" the input/output file sets. If you want, we could take a look at this together at some point and see whether we can do better.

Reply:
I have removed the --sort for now.

This PR is now a lot smaller to fix obvious bugs and add some minor doc improvements.

PTAL.